executor/runcexecutor/executor_linux.go

```diff
@@ -0,0 +1,308 @@
+		NoPivot: w.noPivot,
+		PidFile: pidfile.Name(),
+		Detach:  true,
+	}
+
+	initErrCh := make(chan error)
+	if process.Meta.Tty {
+		var socket runc.ConsoleSocket
+		var cleanup func()
+		socket, cleanup, err = w.setupConsole(ctx, initErrCh, ended, bundle, id, process)
+		if err != nil {
+			return containerd.UnknownExitStatus, err
+		}
+		defer cleanup()
+		createOpts.ConsoleSocket = socket
+	} else {
+		close(initErrCh)
+		createOpts.IO = &forwardIO{stdin: process.Stdin, stdout: process.Stdout, stderr: process.Stderr}
+	}
+
+	exitCh := reaper.Default.Subscribe()
+
+	eg, runCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		_, err = w.runc.Run(runCtx, id, bundle, createOpts)
+		return err
+	})
+	eg.Go(func() error {
+		return <-initErrCh
+	})
+	err = eg.Wait()
+	if err != nil {
+		return containerd.UnknownExitStatus, errors.Wrap(err, "runc create failed")
+	}
+
+	pid, err := runc.ReadPidFile(pidfile.Name())
+	if err != nil {
+		return containerd.UnknownExitStatus, errors.Wrap(err, "failed to read runc pidfile")
+	}
+
+	status := waitpid(ctx, pid, exitCh)
+	defer endedOnce.Do(func() {
+		close(ended)
+	})
+	return status, nil
+}
+
+func (w *runcExecutor) exec(ctx context.Context, id, bundle string, pid0 int, specsProcess *specs.Process, process executor.ProcessInfo) (int, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	pidfile, err := ioutil.TempFile(bundle, "exec-*.pid")
+	if err != nil {
+		return containerd.UnknownExitStatus, errors.Wrap(err, "failed to create pidfile")
+	}
+	defer os.Remove(pidfile.Name())
+	pidfile.Close()
+
+	execOpts := &runc.ExecOpts{
+		PidFile: pidfile.Name(),
+		Detach:  true,
+	}
+	initErrCh := make(chan error)
+	if process.Meta.Tty {
+		socket, cleanup, err := w.setupConsole(ctx, initErrCh, nil, bundle, id, process)
+		if err != nil {
+			return containerd.UnknownExitStatus, errors.Wrap(err, "failed to setup console")
+		}
+		defer cleanup()
+		execOpts.ConsoleSocket = socket
+	} else {
+		close(initErrCh)
+		execOpts.IO = &forwardIO{stdin: process.Stdin, stdout: process.Stdout, stderr: process.Stderr}
+	}
+	exitCh := reaper.Default.Subscribe()
+	eg, execCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		err = w.runc.Exec(execCtx, id, *specsProcess, execOpts)
+		return err
+	})
+	eg.Go(func() error {
+		return <-initErrCh
+	})
+	err = eg.Wait()
+	if err != nil {
+		return containerd.UnknownExitStatus, errors.Wrap(err, "runc create failed")
+	}
+
+	pid, err := runc.ReadPidFile(pidfile.Name())
+	if err != nil {
+		return containerd.UnknownExitStatus, errors.Wrap(err, "failed to read runc pidfile")
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// watch for parent process to die, if that happens then cancel the
+		// exec context
+		exitCh := reaper.Default.Subscribe()
+		waitpid(ctx, pid0, exitCh)
+		cancel()
+	}()
+	defer wg.Wait()
+
+	status := waitpid(ctx, pid, exitCh)
+	cancel() // stop the parent waitpid if still running
```

Comment:
Why do we need this `cancel` when we already have a `defer cancel()` on line 124? AFAIK, this call isn't blocking.

Reply:
We need to cancel so the parent waitpid will exit:
```
	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()
		// watch for parent process to die, if that happens then cancel the
		// exec context
		exitCh := reaper.Default.Subscribe()
		waitpid(ctx, pid0, exitCh)
		cancel()
	}()
	defer wg.Wait()
```
The `defer wg.Wait()` will hang until `cancel()` is called, or the parent exits.  In the case that the parent continues to run we have to manually call `cancel()` so the `defer wg.Wait()` will exit.  The previous `defer cancel()` is just to clean up resources leak in the case of error.